File: pipeline/reimpl/cbtrn03c.py
"""
Reimplementation of CBTRN03C — CardDemo Transaction Detail Report Generator.

Function: Read transaction file sequentially, filter by date range, look up
cross-reference / transaction type / category, and produce a paginated report.

Report is broken by card number with page totals, account totals, and grand total.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from decimal import Decimal
from typing import Callable

from .carddemo_data import TranRecord, CardXrefRecord


# ── Supporting reference structures ───────────────────────────────────────────

@dataclass
class TranTypeRecord:
    tran_type: str = ""    # PIC X(02)
    tran_type_desc: str = ""  # PIC X(58)


@dataclass
class TranCatgRecord:
    tran_type_cd: str = ""
    tran_cat_cd: int = 0
    tran_catg_desc: str = ""  # PIC X(54)


# ── Report line structures ─────────────────────────────────────────────────────

@dataclass
class ReportLine:
    content: str
    is_header: bool = False
    is_total: bool = False
    is_page_break: bool = False


# ── Result ─────────────────────────────────────────────────────────────────────

@dataclass
class ReportResult:
    report_lines: list[ReportLine] = field(default_factory=list)
    records_read: int = 0
    records_reported: int = 0
    grand_total: Decimal = Decimal("0.00")
    page_count: int = 0


PAGE_SIZE = 20


# ── Repository interfaces ──────────────────────────────────────────────────────

class XrefRepository:
    def __init__(self, xrefs: dict[str, CardXrefRecord]):
        self._by_card = xrefs

    def find(self, card_num: str) -> CardXrefRecord | None:
        return self._by_card.get(card_num)


class TranTypeRepository:
    def __init__(self, types: dict[str, TranTypeRecord]):
        self._types = types

    def find(self, type_cd: str) -> TranTypeRecord | None:
        return self._types.get(type_cd)


class TranCatgRepository:
    def __init__(self, catgs: dict[tuple[str, int], TranCatgRecord]):
        self._catgs = catgs

    def find(self, type_cd: str, cat_cd: int) -> TranCatgRecord | None:
        return self._catgs.get((type_cd, cat_cd))


# ── Core processing ───────────────────────────────────────────────────────────

def generate_transaction_report(
    transactions: list[TranRecord],
    xref_repo: XrefRepository,
    trantype_repo: TranTypeRepository,
    trancatg_repo: TranCatgRepository,
    start_date: str,
    end_date: str,
    logger: Callable[[str], None] = print,
) -> ReportResult:
    """Generate transaction detail report — mirrors CBTRN03C PROCEDURE DIVISION."""
    result = ReportResult()
    logger("START OF EXECUTION OF PROGRAM CBTRN03C")

    line_counter = 0
    page_total = Decimal("0.00")
    account_total = Decimal("0.00")
    grand_total = Decimal("0.00")
    curr_card_num = ""
    first_time = True

    def emit(line: str, is_header=False, is_total=False):
        result.report_lines.append(ReportLine(line, is_header, is_total))

    def write_page_header(page_num: int):
        emit("=" * 133, is_header=True)
        emit(f"  CARDEMO TRANSACTION DETAIL REPORT               PAGE: {page_num:05d}", is_header=True)
        emit(f"  DATE RANGE: {start_date} TO {end_date}", is_header=True)
        emit("=" * 133, is_header=True)
        emit(f"{'TRAN ID':<16} {'TRAN DETAILS':<51} {'TRAN AMOUNT':>13}", is_header=True)
        emit("-" * 133, is_header=True)
        result.page_count += 1

    def write_account_totals(card_num: str, acc_total: Decimal):
        emit("-" * 133, is_total=True)
        emit(f"  ACCOUNT TOTAL FOR CARD {card_num}: ${acc_total:>14.2f}", is_total=True)
        emit("", is_total=True)

    write_page_header(1)

    for tran in transactions:
        result.records_read += 1

        # Date filter
        proc_date = tran.tran_proc_ts[:10] if tran.tran_proc_ts else ""
        if proc_date < start_date or proc_date > end_date:
            continue

        result.records_reported += 1

        # Account break
        if tran.tran_card_num != curr_card_num:
            if not first_time:
                write_account_totals(curr_card_num, account_total)
                account_total = Decimal("0.00")
            else:
                first_time = False
            curr_card_num = tran.tran_card_num

        # Look up references
        xref = xref_repo.find(tran.tran_card_num)
        tran_type = trantype_repo.find(tran.tran_type_cd)
        tran_catg = trancatg_repo.find(tran.tran_type_cd, tran.tran_cat_cd)

        type_desc = tran_type.tran_type_desc.strip() if tran_type else tran.tran_type_cd
        catg_desc = tran_catg.tran_catg_desc.strip() if tran_catg else str(tran.tran_cat_cd)
        detail = f"{type_desc}/{catg_desc} {tran.tran_desc[:30]}"

        line = f"{tran.tran_id:<16} {detail:<51} ${tran.tran_amt:>12.2f}"
        emit(line)
        logger(str(tran))

        page_total += tran.tran_amt
        account_total += tran.tran_amt
        grand_total += tran.tran_amt
        line_counter += 1

        # Page break
        if line_counter >= PAGE_SIZE:
            emit(f"  PAGE TOTAL: ${page_total:>14.2f}", is_total=True)
            grand_total_running = grand_total
            emit(f"  GRAND TOTAL SO FAR: ${grand_total_running:>14.2f}", is_total=True)
            page_total = Decimal("0.00")
            line_counter = 0
            write_page_header(result.page_count + 1)

    # Final account total
    if not first_time:
        write_account_totals(curr_card_num, account_total)

    # Grand total
    emit("=" * 133, is_total=True)
    emit(f"  GRAND TOTAL: ${grand_total:>14.2f}", is_total=True)
    emit("=" * 133, is_total=True)

    result.grand_total = grand_total
    logger(f"TRAN-AMT {grand_total}")
    logger(f"WS-PAGE-TOTAL {page_total}")
    logger("END OF EXECUTION OF PROGRAM CBTRN03C")
    return result

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cbtrn03c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Java reimplementation of CardDemo batch program CBTRN03C — Transaction Detail Report.
 *
 * <p>Mirrors {@code pipeline/reimpl/cbtrn03c.py} byte-for-byte. Python is the
 * source of truth.
 */
public class Cbtrn03c implements ProgramRunner {

    private static final BigDecimal BD_0_00 = new BigDecimal("0.00");
    private static final int PAGE_SIZE = 20;

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenarioName = inputs.getOrDefault("SCENARIO", "PROCESS_RECORDS").toUpperCase();
        Scenario scn = buildScenario(scenarioName);
        if (scn == null) {
            Map<String, String> err = new LinkedHashMap<>();
            err.put("error", "unknown scenario: '" + scenarioName + "'");
            return err;
        }

        ReportResult result = generateReport(scn);

        Map<String, String> out = new LinkedHashMap<>();
        out.put("RECORDS_READ", String.valueOf(result.recordsRead));
        out.put("RECORDS_REPORTED", String.valueOf(result.recordsReported));
        out.put("GRAND_TOTAL", result.grandTotal.setScale(2, RoundingMode.HALF_UP).toPlainString());
        out.put("PAGE_COUNT", String.valueOf(result.pageCount));
        for (int i = 0; i < result.reportLines.size(); i++) {
            out.put("REPORT_LINE_" + i, result.reportLines.get(i));
        }
        return out;
    }

    private ReportResult generateReport(Scenario scn) {
        ReportResult result = new ReportResult();
        int lineCounter = 0;
        BigDecimal pageTotal = BD_0_00;
        BigDecimal accountTotal = BD_0_00;
        BigDecimal grandTotal = BD_0_00;
        String currCardNum = "";
        boolean firstTime = true;

        // Write first page header
        writePageHeader(result, 1, scn.startDate, scn.endDate);

        for (TranRecord tran : scn.transactions) {
            result.recordsRead++;

            // Date filter
            String procDate = tran.tranProcTs.length() >= 10
                    ? tran.tranProcTs.substring(0, 10) : "";
            if (procDate.compareTo(scn.startDate) < 0 || procDate.compareTo(scn.endDate) > 0) {
                continue;
            }

            result.recordsReported++;

            // Account break
            if (!tran.tranCardNum.equals(currCardNum)) {
                if (!firstTime) {
                    writeAccountTotals(result, currCardNum, accountTotal);
                    accountTotal = BD_0_00;
                } else {
                    firstTime = false;
                }
                currCardNum = tran.tranCardNum;
            }

            // Look up references
            TranTypeRecord tranType = scn.tranTypes.get(tran.tranTypeCd);
            String typeDesc = tranType != null ? tranType.tranTypeDesc.trim() : tran.tranTypeCd;
            TranCatgKey catgKey = new TranCatgKey(tran.tranTypeCd, tran.tranCatCd);
            TranCatgRecord tranCatg = scn.tranCatgs.get(catgKey);
            String catgDesc = tranCatg != null ? tranCatg.tranCatgDesc.trim() : String.valueOf(tran.tranCatCd);
            String detail = typeDesc + "/" + catgDesc + " " + (tran.tranDesc.length() > 30
                    ? tran.tranDesc.substring(0, 30) : tran.tranDesc);

            String line = String.format("%-16s %-51s $%12.2f",
                    tran.tranId, detail, tran.tranAmt.doubleValue());
            result.reportLines.add(line);

            pageTotal = pageTotal.add(tran.tranAmt);
            accountTotal = accountTotal.add(tran.tranAmt);
            grandTotal = grandTotal.add(tran.tranAmt);
            lineCounter++;

            // Page break
            if (lineCounter >= PAGE_SIZE) {
                result.reportLines.add(String.format("  PAGE TOTAL: $%14.2f", pageTotal.doubleValue()));
                result.reportLines.add(String.format("  GRAND TOTAL SO FAR: $%14.2f", grandTotal.doubleValue()));
                pageTotal = BD_0_00;
                lineCounter = 0;
                writePageHeader(result, result.pageCount + 1, scn.startDate, scn.endDate);
            }
        }

        // Final account total
        if (!firstTime) {
            writeAccountTotals(result, currCardNum, accountTotal);
        }

        // Grand total
        StringBuilder eq = new StringBuilder();
        for (int i = 0; i < 133; i++) eq.append('=');
        result.reportLines.add(eq.toString());
        result.reportLines.add(String.format("  GRAND TOTAL: $%14.2f", grandTotal.doubleValue()));
        result.reportLines.add(eq.toString());

        result.grandTotal = grandTotal;
        return result;
    }

    private void writePageHeader(ReportResult result, int pageNum, String startDate, String endDate) {
        StringBuilder eq = new StringBuilder();
        for (int i = 0; i < 133; i++) eq.append('=');
        StringBuilder dash = new StringBuilder();
        for (int i = 0; i < 133; i++) dash.append('-');

        result.reportLines.add(eq.toString());
        result.reportLines.add(String.format("  CARDEMO TRANSACTION DETAIL REPORT               PAGE: %05d", pageNum));
        result.reportLines.add("  DATE RANGE: " + startDate + " TO " + endDate);
        result.reportLines.add(eq.toString());
        result.reportLines.add(String.format("%-16s %-51s %13s", "TRAN ID", "TRAN DETAILS", "TRAN AMOUNT"));
        result.reportLines.add(dash.toString());
        result.pageCount++;
    }

    private void writeAccountTotals(ReportResult result, String cardNum, BigDecimal accTotal) {
        StringBuilder dash = new StringBuilder();
        for (int i = 0; i < 133; i++) dash.append('-');
        result.reportLines.add(dash.toString());
        result.reportLines.add(String.format("  ACCOUNT TOTAL FOR CARD %s: $%14.2f", cardNum, accTotal.doubleValue()));
        result.reportLines.add("");
    }

    // ── Scenarios ─────────────────────────────────────────────────────────

    private Scenario buildScenario(String name) {
        switch (name) {
            case "PROCESS_RECORDS": return scenarioProcessRecords();
            case "EMPTY_INPUT":    return scenarioEmptyInput();
            default: return null;
        }
    }

    private Scenario scenarioProcessRecords() {
        Scenario s = new Scenario();
        s.startDate = "2026-04-01";
        s.endDate = "2026-04-30";

        s.tranTypes.put("01", new TranTypeRecord("01", "Purchase"));
        s.tranCatgs.put(new TranCatgKey("01", 1),
                new TranCatgRecord("01", 1, "Retail"));

        s.transactions.add(new TranRecord(
                "TRN0000000000001", "01", 1, "Widget purchase",
                new BigDecimal("150.25"), "4111000000001111",
                "2026-04-01-12.00.00.000000"));
        s.transactions.add(new TranRecord(
                "TRN0000000000002", "01", 1, "Gadget purchase",
                new BigDecimal("75.50"), "4111000000001111",
                "2026-04-05-14.30.00.000000"));
        return s;
    }

    private Scenario scenarioEmptyInput() {
        Scenario s = new Scenario();
        s.startDate = "2026-04-01";
        s.endDate = "2026-04-30";
        return s;
    }

    // ── Inner data types ─────────────────────────────────────────────────

    private static class Scenario {
        String startDate = "";
        String endDate = "";
        final Map<String, TranTypeRecord> tranTypes = new HashMap<>();
        final Map<TranCatgKey, TranCatgRecord> tranCatgs = new HashMap<>();
        final List<TranRecord> transactions = new ArrayList<>();
    }

    static class TranRecord {
        final String tranId;
        final String tranTypeCd;
        final int tranCatCd;
        final String tranDesc;
        final BigDecimal tranAmt;
        final String tranCardNum;
        final String tranProcTs;

        TranRecord(String tranId, String tranTypeCd, int tranCatCd,
                   String tranDesc, BigDecimal tranAmt, String tranCardNum,
                   String tranProcTs) {
            this.tranId = tranId;
            this.tranTypeCd = tranTypeCd;
            this.tranCatCd = tranCatCd;
            this.tranDesc = tranDesc;
            this.tranAmt = tranAmt;
            this.tranCardNum = tranCardNum;
            this.tranProcTs = tranProcTs;
        }
    }

    static class TranTypeRecord {
        final String tranType;
        final String tranTypeDesc;

        TranTypeRecord(String tranType, String tranTypeDesc) {
            this.tranType = tranType;
            this.tranTypeDesc = tranTypeDesc;
        }
    }

    static class TranCatgKey {
        final String typeCd;
        final int catCd;

        TranCatgKey(String typeCd, int catCd) {
            this.typeCd = typeCd;
            this.catCd = catCd;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof TranCatgKey)) return false;
            TranCatgKey k = (TranCatgKey) o;
            return catCd == k.catCd && typeCd.equals(k.typeCd);
        }

        @Override
        public int hashCode() {
            return 31 * typeCd.hashCode() + catCd;
        }
    }

    static class TranCatgRecord {
        final String tranTypeCd;
        final int tranCatCd;
        final String tranCatgDesc;

        TranCatgRecord(String typeCd, int catCd, String catgDesc) {
            this.tranTypeCd = typeCd;
            this.tranCatCd = catCd;
            this.tranCatgDesc = catgDesc;
        }
    }

    static class ReportResult {
        final List<String> reportLines = new ArrayList<>();
        int recordsRead = 0;
        int recordsReported = 0;
        BigDecimal grandTotal = BD_0_00;
        int pageCount = 0;
    }
}
